File: C++/src/AoIQueue.cpp
#include "AoIQueue.h"


AoIQueue::AoIQueue() {
    init();
}


AoIQueue::AoIQueue(double end, double dt, double lam, double m) {
    init();

    // Assign member variables
    _tFinal = end;
    _tStep = dt;
    _lambda = lam;
    _mu = m;

    // Generate number of packet arrivals
    _nEvents = 0;
    while (_nEvents == 0) {
        GenerateNumEvents();
    }
    

    // Generate arrival times for each packet
    GenerateArrivals();

    // Generate service times for each packet
    GenerateServiceTimes();

    // Calculate when each packet is finished being served
    CalculatePacketServed();
   
    // Cut of simulation time to when the last packet was served. This might
    // balance simulations where packets stop being served early on in the
    // simulation
    _tFinal = _timeFinished[_nEvents-1];

    // Calculate the amount of steps in the simulation
    _nIntervals = (_tFinal / _tStep) + 1;

    CalculateAge();
}

AoIQueue::~AoIQueue() {
    clear();
}

// Initializes member variables
void AoIQueue::init() {
    // Initialize pointers to NULL
    _timeArrived = NULL;
    _timeFinished = NULL;
    _packetAge = NULL;
    _serviceTime = NULL;
    _delayTime = NULL;
    _age = NULL;
    _time = NULL;
}

void AoIQueue::clear() {
    // Delete all dynamically allocated arrays
    if (_timeArrived != NULL) { delete [] _timeArrived; }
    if (_timeFinished != NULL) { delete [] _timeFinished; }
    if (_packetAge != NULL) { delete [] _packetAge; }
    if (_serviceTime != NULL) { delete [] _serviceTime; }
    if (_delayTime != NULL) { delete [] _delayTime; }
    if (_age != NULL) { delete [] _age; }
    if (_time != NULL) { delete [] _time; }
}

void AoIQueue::allocateArrays() {
    // Allocate arrays based on number of arrivals
    _timeArrived = new double[_nEvents];
    _timeFinished = new double[_nEvents];
    _packetAge = new double[_nEvents];
    _serviceTime = new double[_nEvents];
    _delayTime = new double[_nEvents];
}

void AoIQueue::GenerateNumEvents() {
    // Expected number of events based on lambda and total simulation time
    int expectedEvents = _tFinal * _lambda;

    // Create random number generator
    std::random_device rd;
    std::mt19937 generator (rd());

    std::poisson_distribution<int> poissonDist(expectedEvents);

    // Generate number of transmission events
    _nEvents = poissonDist(generator);

    // Allocate all arrays that are of size _nEvents
    allocateArrays();
}

void AoIQueue::GenerateArrivals() {
    // Create random number generator based off exponential distribution 
    std::random_device rd;
    std::mt19937 generator (rd());

    std::exponential_distribution<double> expDist(_lambda);

    // Generate timestamps of arrivals for each packet
    for (int i = 0; i < _nEvents; ++i) {
        // Generate time since last packet using exponential distribution
        double timestamp = expDist(generator);
        // Round to match step size precision
        timestamp = round(timestamp / _tStep) * _tStep;

        // Add to arrivals vector
        if (i == 0){ // First packet
            _timeArrived[i] = timestamp;
        }
        else { // Add up the times to get the timestamp an arrival happened
            _timeArrived[i] = timestamp + _timeArrived[i-1];
        }
    }
}

void AoIQueue::GenerateServiceTimes() {
    // Create random number generator based off exponential distribution 
    std::random_device rd;
    std::mt19937 generator (rd());

    std::exponential_distribution<double> expDist(_mu);

    // Generate timestamps of arrivals for each packet
    for (int i = 0; i < _nEvents; ++i) {
        // Generate time since last packet using exponential distribution
        double timestamp = expDist(generator);

        // Round to match step size precision
        timestamp = round(timestamp / _tStep) * _tStep;

        // Add to arrivals vector
        _serviceTime[i] = timestamp;
    }
}

void AoIQueue::CalculatePacketServed() {
    // Calculate time first packet finished being served
    _delayTime[0] = 0;
    _timeFinished[0] = _timeArrived[0] + _delayTime[0] + _serviceTime[0];
    _packetAge[0] = _timeFinished[0] - _timeArrived[0];

    // Sum of all delay times to calculate average delay
    double sumDelay = 0;

    for (int i = 1; i < _nEvents; ++i) {
        if (_timeArrived[i] >= _timeFinished[i-1]) {
            _delayTime[i] = 0;
        }
        else {
            _delayTime[i] = _timeFinished[i-1] - _timeArrived[i];
        }

        sumDelay += _delayTime[i];
        _timeFinished[i] = _timeArrived[i] + _delayTime[i] + _serviceTime[i];
        _packetAge[i] = _timeFinished[i] - _timeArrived[i];
    }

    // Store the average delay
    _avgDelay = sumDelay / _nEvents;
}

void AoIQueue::CalculateAge() {
    // Allocate age and time arrays
    _age = new double[_nIntervals];
    _time = new double[_nIntervals];

    // Set the initial age
    _age[0] = 0;
    double sumAge = 0;

    // Pointers to keep track of the current packet
    double* currentPacket = _timeFinished;
    double* currentAge = _packetAge;

    // Iterate through each time step
    for (int i = 0; i < _nIntervals; ++i) {
        // Record the current time
        _time[i] = i * _tStep;

        // Increase the age by _tStep
        if (i > 0) {
            _age[i] = _age[i-1] + _tStep;
        }

        // Check if a packet is done being served at the current time
        double timeToPacket = _time[i] - *currentPacket;
        double cutoff = _tStep / 10;
        if ( (timeToPacket < cutoff) && (timeToPacket > -cutoff)) {
            // Decrease the age at current time to the age of the new packet
            _age[i] = *currentAge;

            // Increment the pointers
            currentPacket++;
            currentAge++;
        }

        sumAge += _age[i];
    }

    // Store the avgerage age
    _avgAge = sumAge / _nIntervals;
}

void AoIQueue::print() {
    std::cout << "Total number of events: " << _nEvents << std::endl;
    std::cout << "lambda: " << _lambda << std::endl;
    std::cout << "    mu: " << _mu << std::endl;
    // Print data
    std::cout << std::endl;
    std::cout << std::setw(10) << "Arrival";
    std::cout << std::setw(10) << "Service";
    std::cout << std::setw(10) << "Delay";
    std::cout << std::setw(10) << "Served";
    std::cout << std::setw(10) << "Age";
    
    std::cout << std::endl;
    for (int i = 0; i < _nEvents; ++i) {
        std::cout << std::setw(10) << _timeArrived[i];
        std::cout << std::setw(10) << _serviceTime[i];
        std::cout << std::setw(10) << _delayTime[i];
        std::cout << std::setw(10) << _timeFinished[i];
        std::cout << std::setw(10) << _packetAge[i];
        std::cout << std::endl;
    }

    std::cout << std::endl;

    std::cout << "Average Age = " << _avgAge << std::endl;
    std::cout << "Average Delay = " << _avgDelay << std::endl;
    std::cout << std::endl;
}

bool AoIQueue::exportData(const std::string& filename) {
    std::ofstream outFile(filename);
    if (!outFile.good()) {
        std::cerr << "Could not open " << filename << " to write" << std::endl;
        return false;
    }

    outFile << "time,age,avgAge" <<std::endl;

    for (int i = 0; i < _nIntervals; ++i) {
        outFile << _time[i] << "," << _age[i] << "," << _avgAge << std::endl;
    }

    return true;
}

File: Python/AoIQueue.py
import numpy as np
import matplotlib.pyplot as plt 
from helper import GenerateTransmissions, GenerateServiceTime
import math

class AoIQueue:
    def __init__(self, tFinal, tStep, arrivalRate, serviceRate):
        # Initialize simulation parameters
        self._tFinal = tFinal  # Total simulation time
        self._tStep = tStep  # Simulation step size
        self._lambda = arrivalRate  # Packet Arrival Rate (packets/second)
        self._mu = serviceRate  # Service Rate (packets served/second)

        # Decimal places to round to based on step size
        precision = int(-math.log10(self._tStep))

        # Make time array
        padding = 5 * 1/self._lambda  # Add some time to the begining to average 10 packet arrivals 
        self._t = np.arange(0, padding + self._tFinal + self._tStep, self._tStep)
        start_time = padding
        end_time = max(self._t) #- padding

        # Generate array of packet arrival times and store the number of packets
        self._timeArrived = GenerateTransmissions(self._t, self._lambda)
          

        # Generate array of service times
        array_size = self._timeArrived.shape
        self._serviceTimes = GenerateServiceTime(
            self._mu, self._tStep,  array_size)

        # Make array for wait times in queue. The first value is 0
        self._delayTime = np.zeros(array_size)
        self._queueWait = np.zeros(array_size)
        # Make array to keep track of when packets finish being served
        self._timeFinished = np.zeros(array_size)
        # Packet age array to keep track of the age of each packet
        self._packetAge = np.zeros(array_size)

        # Initialize age array with initial age of 0
        self._age = np.copy(self._t)
        numServed = 0  # Counter for number of packets served

        # Go through each packet and calculate when they get served
        for i in range(len(self._timeArrived)):
            # Calculate when the first packet gets served
            if i == 0:
                # First packet that arrives is only as old as its service time
                firstPacketServed = self._timeArrived[0] + self._serviceTimes[0]
                self._timeFinished[0] = firstPacketServed
                self._packetAge[0] = self._serviceTimes[0]

            # Calculate the age of the rest of the packets
            else:
                if self._timeArrived[i] >= self._timeFinished[i-1]:
                    # Current packet arrived after last packed was finished serving
                    self._queueWait[i] = 0  # It doesn't have to wait

                else:
                    # Packet arrived before last packet is finished serving
                    self._queueWait[i] = self._timeFinished[i-1] - self._timeArrived[i]

                # Calculate when current packet gets served
                packetServed = self._timeArrived[i] + self._queueWait[i] + self._serviceTimes[i]
                self._timeFinished[i] = packetServed

                currentPacketAge = packetServed - self._timeArrived[i]
                self._packetAge[i] = currentPacketAge

            # Update the age for each packet
        
            # Get the time that current packet finished serving
            currentTime = self._timeFinished[i]  
            currentTime = round(currentTime, precision)

            # Cutoff the simulation at tFinal
            if currentTime > end_time:
                break

            # Find the corresponding index in the age array
            ageIndex = currentTime / self._tStep
            ageIndex = int(round(ageIndex))

            # Check for rounding errors
            if abs(self._t[ageIndex] - currentTime) > 1e-5:
                print(currentTime, end=' ')
                print(ageIndex, end=' ')
                print(self._t[ageIndex])
            
            # Decrease the age to the age of the packet that just got served
            reduceAge = self._age[ageIndex] - self._packetAge[i]
            self._age[ageIndex:] -= reduceAge

            if currentTime >= start_time and currentTime <= end_time:
                numServed += 1


        # Cutoff the padding at the beginning
        self._t = np.arange(0, self._tFinal+self._tStep, self._tStep)
        start_idx = int( start_time / self._tStep )
        end_idx = int( end_time / self._tStep ) + 1
        self._age = self._age[start_idx:end_idx] 


        # Calculate the averages
        self.avgAge = np.mean(self._age)
        self.avgQueueWait = np.mean(self._queueWait) 

        self._delayTime = self._queueWait + self._serviceTimes
        self.avgDelay = np.mean(self._delayTime)

        # Expected number of packets for the time window
        packets_expected = self._tFinal * self._lambda
        self.percentServed = numServed / packets_expected


    # Prints information about the simulation
    def printData(self):
        print("Simulating to time {:.1f}s with a step size of {:.2f}".format(
            self._tFinal, self._tStep))
        print("lambda: {:.3f} packets arriving per second".format(self._lambda))
        print("mu: {:.3f} packets served per second".format(self._mu))
        
        print("Arrival".rjust(10), end='')
        print("Service".rjust(10), end='')
        print("Delay".rjust(10), end='')
        print("Finished".rjust(10), end='')
        print("Age".rjust(10))

        for i in range(len(self._timeArrived)):
            print("{:.1f}".format(self._timeArrived[i]).rjust(10), end='')
            print("{:.1f}".format(self._serviceTimes[i]).rjust(10), end='')
            print("{:.1f}".format(self._delayTime[i]).rjust(10), end='')
            print("{:.1f}".format(self._timeFinished[i]).rjust(10), end='')
            print("{:.1f}".format(self._packetAge[i]).rjust(10))

    def plotAge(self):
        fig, ax = plt.subplots()
        ax.plot(self._t, self._age, label="Age of Information")
        avgAgePlt = self.avgAge * np.ones( np.size(self._age) )
        ax.plot(self._t, avgAgePlt,
                 label="Average Age = {:.2f}".format(self.avgAge))
        ax.set_xlabel("time (s)")
        ax.set_ylabel("age (s)")
        ax.legend()
        
      
    # Exports the age and time arrays in a csv file for other programs
    def exportResults(self, filename):
        outFile = open(filename, "w")
        outFile.write("time,age\n")
        for i in range(len(self._age)):
            outFile.write("{:f},{:f}\n".format(self._t[i], self._age[i]))

    # Exports timeTransmit and serviceTimes as csv file.
    # Mostly to use with matlab to test that the algorithm is correct
    def exportSimulationParams(self, filename):
        outFile = open(filename, "w")
        outFile.write("timeArrived, serviceTimes\n")
        for i in range(len(self._timeArrived)):
            outFile.write("{:f},{:f}\n".format(self._timeArrived[i], self._serviceTimes[i]))

    # # Returns the average age of the simulation
    # def getAvgAge(self):
    #     return self._avgAge

    # # Returns the average delay of the simulation
    # def getAvgDelay(self):
    #     return self._avgDelay

    # def getPercentServed(self):
    #     return self._percentServed

if __name__ == "__main__":
    tFinal = 18000
    dt = 0.1
    arrivalRate = 1/200
    serviceRate = 1/30

    
    queue = AoIQueue(tFinal, dt, arrivalRate, serviceRate)
    

    avgAges = queue.avgAge
    queue.printData()
    print(queue.percentServed)

    queue.plotAge()
    plt.show()
    
    

